File: src/services/audit/detectors/ad/advanced/laps/laps-not-deployed.ts
/**
 * LAPS Not Deployed Detector
 * Check for LAPS not deployed
 */

import { ADComputer } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';

export function detectLapsNotDeployed(computers: ADComputer[], includeDetails: boolean): Finding {
  const affected = computers.filter((c) => {
    return !('ms-Mcs-AdmPwd' in c) && !('msLAPS-Password' in c);
  });

  return {
    type: 'LAPS_NOT_DEPLOYED',
    severity: 'medium',
    category: 'advanced',
    title: 'LAPS Not Deployed',
    description: 'LAPS not deployed on domain computers. Shared/static local admin passwords across workstations.',
    count: affected.length,
    affectedEntities: includeDetails ? affected.map((c) => c.dn) : undefined,
  };
}

File: internal/audit/detectors/ad/advanced/laps/laps-not-deployed.go
package laps

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// LapsNotDeployedDetector detects computers without LAPS
type LapsNotDeployedDetector struct {
	audit.BaseDetector
}

// NewLapsNotDeployedDetector creates a new detector
func NewLapsNotDeployedDetector() *LapsNotDeployedDetector {
	return &LapsNotDeployedDetector{
		BaseDetector: audit.NewBaseDetector("LAPS_NOT_DEPLOYED", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *LapsNotDeployedDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.Computer

	for _, c := range data.Computers {
		// No LAPS (neither legacy nor Windows LAPS)
		if c.LegacyLAPSPassword == "" && c.WindowsLAPSPassword == "" {
			affected = append(affected, c)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "LAPS Not Deployed",
		Description: "LAPS not deployed on domain computers. Shared/static local admin passwords across workstations.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedComputerEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewLapsNotDeployedDetector())
}
